cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "groupby.hpp"
+
+#include <copying/scatter.hpp>
+#include <table/legacy/device_table.cuh>
+#include <table/legacy/device_table_row_operators.cuh>
+#include <bitmask/legacy/bit_mask.cuh>
+#include <utilities/column_utils.hpp>
+#include <utilities/cuda_utils.hpp>
+
+#include <cudf/copying.hpp>
+
+#include <thrust/scan.h>
+#include <thrust/binary_search.h>
+#include <thrust/unique.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+
+#include <algorithm>
+#include <tuple>
+#include <numeric>
+
+namespace {
+
+template <bool nullable = true>
+struct transform_row_eq_comparator {
+  row_equality_comparator<nullable> cmp;
+  gdf_size_type const* sorted_order;
+
+  CUDA_DEVICE_CALLABLE
+  bool operator() (gdf_size_type lhs, gdf_size_type rhs) {
+    return cmp(sorted_order[lhs], sorted_order[rhs]);
+  }
+};
+
+struct permutation_label_setter {
+  gdf_size_type * group_labels_ptr;
+  gdf_size_type const* group_ids_ptr;
+
+  CUDA_DEVICE_CALLABLE
+  void operator() (gdf_size_type i) { 
+    group_labels_ptr[group_ids_ptr[i]] = 1;
+  }
+};
+
+} // namespace anonymous
+
+
+namespace cudf {
+
+namespace detail {
+
```

Comment:
The `groupby::` methods should probably take an optional stream -- or the constructor of the object should take one. Then, use the stream inside the methods (e.g. change cudaMemset and cudaMemcpy to async, pass it to rmm exec policy, etc.).

Reply:
It's added https://github.com/rapidsai/cudf/pull/2423/files?file-filters%5B%5D=.hpp#diff-f30daf588f8857aa5ded403d4f39ea44R52

Reply:
Ah, I see, you added it as member.